src/witness_generator/machines/block_machine.rs

```diff
@@ -0,0 +1,556 @@
+    }
+
+    fn evaluate(
+        &self,
+        fixed_data: &FixedData,
+        expression: &Expression,
+    ) -> Result<AffineExpression, EvalError> {
+        ExpressionEvaluator::new(SymoblicWitnessEvaluator::new(
+            fixed_data,
+            self.row,
+            WitnessData {
+                fixed_data,
+                data: &self.data,
+                row: self.row,
+            },
+        ))
+        .evaluate(expression)
+    }
+
+    /// Converts a poly ID that migth contain a next offset
+    /// to a regular poly ID plus a row offset.
+    fn convert_next_to_offset(&self, id: usize) -> (usize, u64) {
+        convert_next_to_offset(self.witness_count, id)
+    }
+}
+
+impl BitConstraintSet for BlockMachine {
+    fn bit_constraint(&self, id: usize) -> Option<BitConstraint> {
+        let (poly, offset) = self.convert_next_to_offset(id);
+        self.global_bit_constraints.get(&poly).cloned().or_else(|| {
+            let row = (self.row + offset) % self.degree;
+            self.bit_constraints.get(&poly)?.get(&row).cloned()
+        })
+    }
+}
+
+#[derive(Clone)]
+struct WitnessData<'a> {
+    pub fixed_data: &'a FixedData<'a>,
+    pub data: &'a HashMap<usize, Vec<Option<AbstractNumberType>>>,
+    pub row: DegreeType,
+}
+
+impl<'a> WitnessColumnEvaluator for WitnessData<'a> {
+    fn value(&self, name: &str, next: bool) -> Result<AffineExpression, EvalError> {
+        let id = self.fixed_data.witness_ids[name];
+        let row = if next {
+            (self.row + 1) % self.fixed_data.degree
+        } else {
+            self.row
+        };
+        // It is not an error to access the rows outside the block.
+        // We just return a symbolic ID for those.
+        match self.data[&id].get(row as usize).cloned().flatten() {
+            Some(value) => Ok(value.into()),
+            None => {
+                let witness_count = self.fixed_data.witness_cols.len();
+                let symbolic_id = if next { id + witness_count } else { id };
+                Ok(AffineExpression::from_witness_poly_value(symbolic_id))
+            }
+        }
+    }
+}
+
+impl<'a> WitnessColumnNamer for WitnessData<'a> {
+    fn name(&self, i: usize) -> String {
+        match convert_next_to_offset(self.fixed_data.witness_cols.len(), i) {
+            (i, 0) => self.fixed_data.name(i),
+            (i, 1) => self.fixed_data.name(i) + "\'",
+            _ => panic!(),
+        }
+    }
+}
+
+/// Converts a poly ID that migth contain a next offset
+/// to a regular poly ID plus a row offset.
+fn convert_next_to_offset(witness_count: usize, id: usize) -> (usize, u64) {
+    if id < witness_count {
+        (id, 0)
+    } else {
+        (id - witness_count, 1)
+    }
+}
+
+struct ProcessingSequenceCache {
+    block_size: usize,
+    identities_count: usize,
+    cache: BTreeMap<SequenceCacheKey, Vec<SequenceStep>>,
+}
+
+#[derive(Clone)]
+struct SequenceStep {
+    row_delta: i64,
+    identity_index: Option<usize>,
```

Comment:
What is the `None` case here?

Reply:
Ah it's the same as in `process_identity`. This could be clearer here, maybe with

```
enum Identity {
   Internal(usize),
   Query
}
```